arch/riscv/core/thread.c

```diff
@@ -131,4 +148,216 @@ int arch_float_enable(struct k_thread *thread)
+ *
+ *    3                   2                   1
+ *  1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |                        address[33:2]                          | pmpaddr#
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ *
+ * NAPOT: Each 32-bit pmpaddr# register defines the start address and the size
+ * of the PMP region. The number of concurrent 1s begging at the LSB indicates
+ * the size of the region as a power of two (e.g. 0x...0 = 8-byte, 0x...1 = 
+ * 16-byte, 0x...11 = 32-byte, etc.).
+ *
+ *    3                   2                   1
+ *  1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |                        address[33:2]                |0|1|1|1|1| pmpaddr#
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ *
+ * NA4: This is essentially an edge case of NAPOT where the entire pmpaddr#
+ * register defines a 4-byte wide region.
+ */
+
+int arch_buffer_validate(void * addr, size_t size, int write)
+{
+	size_t pmpcfg;
+	size_t pmpaddr[8];
+	uint8_t rwx[8];
+
+	GET_CSR(pmpcfg0,  pmpcfg);
+	rwx[0] = (pmpcfg >> RV_PMP_0CFG) & RV_PMP_RWX;
+	rwx[1] = (pmpcfg >> RV_PMP_1CFG) & RV_PMP_RWX;
+	rwx[2] = (pmpcfg >> RV_PMP_2CFG) & RV_PMP_RWX;
+	rwx[3] = (pmpcfg >> RV_PMP_3CFG) & RV_PMP_RWX;
+	GET_CSR(pmpcfg1,  pmpcfg);
+	rwx[4] = (pmpcfg >> RV_PMP_0CFG) & RV_PMP_RWX;
+	rwx[5] = (pmpcfg >> RV_PMP_1CFG) & RV_PMP_RWX;
+	rwx[6] = (pmpcfg >> RV_PMP_2CFG) & RV_PMP_RWX;
+	rwx[7] = (pmpcfg >> RV_PMP_3CFG) & RV_PMP_RWX;
+
+	GET_CSR(pmpaddr0, pmpaddr[0]);
+	GET_CSR(pmpaddr1, pmpaddr[1]);
+	GET_CSR(pmpaddr2, pmpaddr[2]);
+	GET_CSR(pmpaddr3, pmpaddr[3]);
+	GET_CSR(pmpaddr4, pmpaddr[4]);
+	GET_CSR(pmpaddr5, pmpaddr[5]);
+	GET_CSR(pmpaddr6, pmpaddr[6]);
+	GET_CSR(pmpaddr7, pmpaddr[7]);
+
+	int i;
+	for (i = 0; i < 8; i++) pmpaddr[i] <<= 2;
+
+	/* pmpcfg0 defines a region from address 0x0 to pmpaddr0 */
+	if ((rwx[0] & RV_PMP_RO) && (((size_t) addr + size) < pmpaddr[0])) {
+		if (write) {
+			if (rwx[0] & RV_PMP_W) return 0;
+			else return 1;
+		} else return 0;
+	}
+	
+	/* pmpcfg1-7 define regions beween the addresses in pmpaddr0-7 */
+	for (i = 1; i < 8; i++) {
+		if ((rwx[i] & RV_PMP_RO) && ((size_t) addr >= pmpaddr[i-1]) 
+				&& (((size_t) addr + size) < pmpaddr[i])) {
+			if (write) {
+				if (rwx[i] & RV_PMP_W) return 0;
+				else return 1;
+			} else return 0;
+		}
+	}
+	 
+	return 1;
+}
+
+int arch_mem_domain_max_partitions_get(void)
+{
+	/*
+	 * RISC-V supports up to sixteen PMP address registers. The current 
+	 * implementation of RISC-V userspace uses TOR PMP settings and uses only
+	 * eight of these registers. Five of these define the bounds of the stack
+	 * for the current thread and the read-only code area. The remaining three
+	 * registers are available to define one user-defined partition. Three 
```

Comment:
This is going to be unusable in practice if memory domains can only have 1 partition. We have ideas about combining library memory into one region https://github.com/zephyrproject-rtos/zephyr/issues/25891 to help with this somewhat but we're not there yet

Reply:
We've re-engineered the implementation a bit to use the NAPOT addressing mode now, so the number of available memory domains has increased from 1 to 5. The other 3 PMP registers are used to protect .text, .rodata and the stack.